torch/optim/lr_scheduler.py

```diff
@@ -0,0 +1,206 @@
+        >>> for epoch in range(10):
+        >>>     train(...)
+        >>>     val_loss = validate(...)
+        >>>     # different from LambdaLR, step should be called after validate()
+        >>>     scheduler.step(epoch, val_loss)
+    """
+
+    def __init__(self, optimizer, mode='min', factor=0.1, patience=10,
+                 verbose=0, threshold=1e-4, threshold_mode='rel',
+                 cooldown=0, min_lr=0):
+
+        if factor >= 1.0:
+            raise ValueError('ReduceLROnPlateau '
+                             'does not support a factor >= 1.0.')
+        self.factor = factor
+        self.min_lr = min_lr
+        self.patience = patience
+        self.verbose = verbose
+        self.cooldown = cooldown
+        self.cooldown_counter = 0  # Cooldown counter.
+        self.monitor_op = _MonitorOp(mode=mode, threshold=threshold,
+                                     threshold_mode=threshold_mode)
+        self.wait = 0
+        self.best = 0
+        assert isinstance(optimizer, Optimizer)
+        self.optimizer = optimizer
+        self._reset()
+
+    def _reset(self):
+        """Resets wait counter and cooldown counter.
+        """
+        self.best = self.monitor_op.worse
+        self.cooldown_counter = 0
+        self.wait = 0
+        self.lr_epsilon = self.min_lr * 1e-4
+
+    def step(self, epoch, metrics):
+        current = metrics
+
+        if self.in_cooldown():
+            self.cooldown_counter -= 1
+            self.wait = 0
+
+        if self.monitor_op(current, self.best):
+            self.best = current
+            self.wait = 0
+        elif not self.in_cooldown():
+            if self.wait >= self.patience:
+                self._reduce_lr(epoch)
+                self.cooldown_counter = self.cooldown
+                self.wait = 0
+            self.wait += 1
+
+    def _reduce_lr(self, epoch):
+        for inx_group, param_group in enumerate(self.optimizer.param_groups, 0):
+            old_lr = float(param_group['lr'])
+            if old_lr > self.min_lr + self.lr_epsilon:
+                new_lr = old_lr * self.factor
+                new_lr = max(new_lr, self.min_lr)
+                param_group['lr'] = new_lr
+                if self.verbose > 0:
+                    print('Epoch %05d: reducing learning rate of group %d to %s.' % (epoch, inx_group, new_lr))
+
+    def in_cooldown(self):
+        return self.cooldown_counter > 0
+
+
+class _MonitorOp(object):
+    def __init__(self, mode, threshold, threshold_mode):
+        if mode not in ['min', 'max']:
+            raise RuntimeError('Learning Rate Plateau Reducing mode %s is unknown!')
+        if threshold_mode not in ['rel', 'abs']:
+            raise RuntimeError('Learning Rate Plateau threshold mode %s is unknown!')
+        if mode == 'min' and threshold_mode == 'rel':
+            rel_epsilon = 1. - threshold
+            self.monitor_op = lambda a, best: np.less(a, best * rel_epsilon)
```

Comment:
You can't use numpy. It's not a hard dependency and the same thing can be easily implemented using only standard Python libraries